// @xmtp/xmtp-js/dist/esm/src/codecs/TypingNotification.js
import { ContentTypeId } from '../MessageContent';
// xmtp.org/typingNotification
//
// This content type is used for typing notifications
export const ContentTypeTypingNotification = new ContentTypeId({
    authorityId: 'xmtp.com',
    typeId: 'typingNotification',
    versionMajor: 1,
    versionMinor: 0,
});
// Important: Typing Notifications should only be sent on ephemeral topics.
export class TypingNotificationCodec {
    get contentType() {
        return ContentTypeTypingNotification;
    }
    encode(content) {
        return {
            type: ContentTypeTypingNotification,
            parameters: {
                timestamp: content.timestamp.toISOString(),
                typerAddress: content.typerAddress,
                isFinished: content.isFinished ? 'true' : 'false',
            },
            content: new Uint8Array(),
        };
    }
    decode(content) {
        const timestamp = new Date(content.parameters.timestamp);
        const typerAddress = content.parameters.senderAddress;
        const isFinished = content.parameters.isFinished === 'true';
        return {
            timestamp,
            typerAddress,
            isFinished,
        };
    }
}
