// @xmtp/xmtp-js/dist/cjs/src/codecs/TypingNotification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypingNotificationCodec = exports.ContentTypeTypingNotification = void 0;
const MessageContent_1 = require("../MessageContent");
// xmtp.org/typingNotification
//
// This content type is used for typing notifications
exports.ContentTypeTypingNotification = new MessageContent_1.ContentTypeId({
    authorityId: 'xmtp.com',
    typeId: 'typingNotification',
    versionMajor: 1,
    versionMinor: 0,
});
// Important: Typing Notifications should only be sent on ephemeral topics.
class TypingNotificationCodec {
    get contentType() {
        return exports.ContentTypeTypingNotification;
    }
    encode(content) {
        return {
            type: exports.ContentTypeTypingNotification,
            parameters: {
                timestamp: content.timestamp.toISOString(),
                typerAddress: content.typerAddress,
                isFinished: content.isFinished ? 'true' : 'false',
            },
            content: new Uint8Array(),
        };
    }
    decode(content) {
        const timestamp = new Date(content.parameters.timestamp);
        const typerAddress = content.parameters.senderAddress;
        const isFinished = content.parameters.isFinished === 'true';
        return {
            timestamp,
            typerAddress,
            isFinished,
        };
    }
}
exports.TypingNotificationCodec = TypingNotificationCodec;
